overlay/usr/lib/inithooks/bin/nextcloud.py

```diff
@@ -27,59 +30,158 @@ def usage(s=None):
+    random_pass = "".join(random.choices(chars, k=36))
+    while True:
+        password = set_password(random_pass)
+        match password[0]:
+            case 0:
+                # success
+                log.write(f"admin random password: {random_pass}", "warn")
+                return
+            case 1:
+                # should occur only in extremely rare cases
+                # trying again should work
+                continue
+            case _:
+                # exceptions and unknown errors
+                log.write(
+                    f"password setting failed - left unset:"
+                    f"\n{password[1]}", "err"
+                )
+                return
+
+
 def main():
+    exit_code = 0
+    opts = []
     try:
-        opts, args = getopt.gnu_getopt(sys.argv[1:], "h",
-                                       ['help', 'pass=', 'domain='])
+        opts, args = getopt.gnu_getopt(sys.argv[1:], "h", ["help",
+                                       "pass=", "domain="])
     except getopt.GetoptError as e:
         usage(e)
 
     password = ""
     domain = ""
     for opt, val in opts:
-        if opt in ('-h', '--help'):
+        if opt in ("-h", "--help"):
             usage()
-        elif opt == '--pass':
+        elif opt == "--pass":
             password = val
-        elif opt == '--domain':
+        elif opt == "--domain":
             domain = val
 
     if not password:
-        d = Dialog('TurnKey GNU/Linux - First boot configuration')
-        prefix = ''
+        message = "Enter new password for the Nextcloud 'admin' account."
+        d = Dialog("TurnKey GNU/Linux - First boot configuration")
         while True:
             password = d.get_password(
                 "Nextcloud Password",
-                prefix + "Enter new password for the Nextcloud 'admin' account.",
-                pass_req=10)
-            try:
-                subprocess.run(
-                         args = ['/usr/local/bin/turnkey-occ', 'user:resetpassword', '--password-from-env', 'admin'],
-                         cwd='/var/www/nextcloud',
-                         env={"OC_PASS": password},
-                         text=True,
-                         capture_output=True,
-                         check=True)
-            except subprocess.CalledProcessError as e:
-                prefix = e.stderr + e.stdout + '\n'
-            else:
+                message,
+                pass_req=10,
+            )
+            assert password is not None
+            match password:
+                case "Random1234":
+                    password = set_random_pass()
+                    exit_code = 1
+                    break
+                case "Skip-password":
+                    log.write("admin password not set - please set manually",
+                              "err")
+                    exit_code = 1
+                    break
+                case _:
+                    new_password = set_password(password)
```

Comment:
We should probably handle the "random password" and "skipping password" cases separately from the password input. Without visual feedback a user might accidentally set their password to "Random123" or "skip-password" by miss-typing these words.

Reply:
Done :grin: